llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -2271,6 +2271,27 @@ Instruction *InstCombinerImpl::foldVectorBinop(BinaryOperator &Inst) {
     }
   }
 
+  // Similar to the combine above, but handles the case for scalable vectors
+  // where both shuffle(V1, 0) and C are splats.
+  //
+  // Op(shuffle(V1, 0), (splat C)) -> shuffle(Op(V1, (splat C)), 0)
+  if (isa<ScalableVectorType>(Inst.getType()) &&
```

Comment:
Code structure wise: Remove the scalable vector check, move this *above* the other transform.  That way fixed vector splats go through this too, and we have a much higher chance of finding any bugs.

Reply:
I originally had this above the other transform but it results in regressions because we lose information about which lanes are poison when splatting. E.g. the original transform will do:

`Op(shuffle(V1, 0), <42, poison>) --> shuffle(Op(V1, <42, poison>), 0)`

But this transform which is less general will do

`Op(shuffle(V1, 0), <42, poison>) --> shuffle(Op(V1, <42, 42>), 0)`

The scalable vector check is an early exit just to avoid calling `getSplatValue` unnecessarily. But if I'm remembering correctly it's actually an invariant too, i.e. no fixed length vectors will actually pass through the new combine because they will have all been caught by the more general combine above.